spec/lib/browse_everything/driver/sharepoint_spec.rb

```diff
@@ -0,0 +1,260 @@
+                'size': 5336,
+                'lastModifiedDateTime': Time.current,
+                'parentReference': {
+                  'driveId': 'drive-id2',
+                  'driveType': 'business',
+                  'id': 'asset-id3'
+                },
+                'file': {
+                  'mimeType': 'application/pdf'
+                },
+                '@microsoft.graph.downloadUrl': 'http://microsoft-download.com'
+              }
+            ]
+          }.to_json
+        end
+
+        before do
+          stub_request(
+            :get, 'https://graph.microsoft.com/v1.0/me/drives/drive-id2/root/children'
+          ).to_return(
+            body: file,
+            status: 200,
+            headers: {}
+          )
+        end
+
+        it 'retrieves files' do
+          expect(contents.first).to be_a BrowseEverything::FileEntry
+          expect(contents.first.location).to eq 'sharepoint:drive-id2/items/asset-id3'
+          expect(contents.first.mtime).to be_a Date
+          expect(contents.first.name).to eq 'asset-name3.pdf'
+          expect(contents.first.size).to eq 5336
+          expect(contents.first.type).to eq 'application/pdf'
+        end
+
+        it 'does not retrieve higher level items' do
+          expect(contents.length).to eq 1
+        end
+      end
+    end
+
+    describe '#link_for' do
+      subject(:link) { driver.link_for('drive-id2/items/asset-id3') }
+      let(:file) do
+        {
+          'id': 'asset-id3',
+          'name': 'asset-name3.pdf',
+          'size': 5336,
+          'lastModifiedDateTime': Time.current,
+          'parentReference': {
+            'driveId': 'drive-id2',
+            'driveType': 'business',
+            'id': 'asset-id3'
+          },
+          'file': {
+            'mimeType': 'application/pdf'
+          },
+          '@microsoft.graph.downloadUrl': 'http://microsoft-download.com/asset-id3'
+        }.to_json
+      end
+
+      before do
+        stub_request(
+          :get, 'https://graph.microsoft.com/v1.0/me/drives/drive-id2/items/asset-id3'
+        ).to_return(
+          body: file,
+          status: 200,
+          headers: {}
+        )
+      end
+
+      it 'generates the link for a Sharepoint asset' do
+        expect(link).to be_an Array
+        expect(link.first).to eq 'http://microsoft-download.com/asset-id3'
+        expect(link.last).to be_a Hash
+        expect(link.last).to include file_name: 'asset-name3.pdf'
+        expect(link.last).to include file_size: 5336
+      end
+    end
+
+    describe '#auth_link' do
+      subject(:uri) { driver.auth_link }
+
+      it 'exposes the authorization endpoint URI' do
+        expect(uri).to be_a Addressable::URI
+        expect(uri.to_s).to eq 'https://login.microsoftonline.com/TENANTID/oauth2/v2.0/authorize?client_id=CLIENTID&scope=offline_access https://graph.microsoft.com/.default&redirect_uri=http://example.com/browse/connect&response_type=code'
```

Comment:
This looks odd.  Are there supposed to be two urls?  From what I can tell Addressable::URI includes the second url in the query string of the first.

Reply:
It is a single URL. Scopes are space separated like `&scope=offline_access%20user.read%20mail.read`. So this URI will be `https://login.microsoftonline.com/TENANTID/oauth2/v2.0/authorize?client_id=CLIENTID&scope=offline_access%20https://graph.microsoft.com/.default&redirect_uri=http://example.com/browse/connect&response_type=code` when it gets encoded. 

The scope pointing to that graph URL is used so that we can assign the required permissions through the microsoft entra admin center rather than having to hardcode the permissions or have them as yet another configuration option in Avalon.

Reply:
:+1:  What do you think about changing this line slightly so that's a little more explicit?
```suggestion
        expect(uri.normalize).to eq 'https://login.microsoftonline.com/TENANTID/oauth2/v2.0/authorize?client_id=CLIENTID&scope=offline_access%20https://graph.microsoft.com/.default&redirect_uri=http://example.com/browse/connect&response_type=code'
```